swarm_dial.go

```diff
@@ -343,23 +344,77 @@ func (s *Swarm) dial(ctx context.Context, p peer.ID) (*Conn, error) {
 	if len(goodAddrs) == 0 {
 		return nil, &DialError{Peer: p, Cause: ErrNoGoodAddresses}
 	}
-	goodAddrsChan := make(chan ma.Multiaddr, len(goodAddrs))
+
+	///////// Partition and rank addresses
+	// Partition into Fd and non-Fd consuming addresses as we first we want
+	// to try dialling ALL Non-Fd consuming addresses.
 	nonBackoff := false
+	var fdAddrs []ma.Multiaddr
+	var nonFdAddrs []ma.Multiaddr
+
 	for _, a := range goodAddrs {
 		// skip addresses in back-off
 		if !s.backf.Backoff(p, a) {
 			nonBackoff = true
-			goodAddrsChan <- a
+			if s.IsFdConsumingAddr(a) {
+				fdAddrs = append(fdAddrs, a)
+			} else {
+				nonFdAddrs = append(nonFdAddrs, a)
+			}
 		}
 	}
-	close(goodAddrsChan)
 	if !nonBackoff {
 		return nil, ErrDialBackoff
 	}
+
+	// sorts addresses in descending order of preference for dialing
+	// local addrs > non-relay public addrs > relay-addrs
+	sortAddrs := func(addrs []ma.Multiaddr) {
+		sort.Slice(addrs, func(i, j int) bool {
+			first := addrs[i]
+			second := addrs[j]
+
+			// first preferred if it is private.
+			if manet.IsPrivateAddr(first) {
+				return true
+			}
+
+			// second preferred if first is non-private and second is private.
+			if manet.IsPrivateAddr(second) {
+				return false
+			}
+
+			// first preferred if first is non-relay and second is not private.
+			if _, err := first.ValueForProtocol(ma.P_CIRCUIT); err != nil {
+				return true
+			}
+
+			// second preferred if first is a relay address.
+			// we don't need this sort to be stable.
+			return false
+		})
+	}
+
+	// Sort/Rank by preference
+	sortAddrs(nonFdAddrs)
+	sortAddrs(fdAddrs)
+
 	/////////
+	// first try to get a connection to a non-fd consuming addr and if that fails,
+	// try to get a connection to a fd consuming addr.
+	var connC transport.CapableConn
+	var errNonFd *DialError
+	var dialErr *DialError
+
+	connC, errNonFd = s.dialAddrs(ctx, p, nonFdAddrs)
```

Comment:
I don't think this is a good idea. QUIC is a UDP-based protocol. UDP is unreliable, it has no connection establishment flow like TCP, the feedback from lack of connectivity/port closed is not immediate like with TCP. If we dialled an IP that was routable but a bad/closed port, I suspect we'd have to wait for the full connection timeout to notice, whereas with TCP we'd notice during the TCP handshake (much sooner). Same applies if firewalls/routers are dropping UDP packets.

What does this buy us? If the ultimate goal is to select a succeeding a QUIC connection over a succeeding TCP connection, sequentialising the dials is going to cause more pain than gain IMO.

Reply:
The TCP handshake is a layer below TCP reliability. It's all IP at that point:

* The feedback for a UDP packet sent to a closed port is the same as the feedback for TCP: an ICMP unreachable packet. Unfortunately, many home firewalls will just _drop_ packets so you'll get no feedback either way.
* If a SYN gets dropped, it will be resent eventually (2s?). I assume QUIC does the same.

The real question is whether or not the ICMP unreachable packet this is exposed to the reader. @marten-seemann, any ideas on this?

> What does this buy us?

It:

* Avoids tying up slots in the file descriptor dial limiter needed by peers that only support TCP.
* Avoids spraying SYN packets (NATs don't like that).

However, it's not critical. I'm fine landing a version of this patch that simply prioritizes UDP connections.

Reply:
> The TCP handshake is a layer below TCP reliability. It's all IP at that point

That is correct.

> If a SYN gets dropped, it will be resent eventually (2s?). I assume QUIC does the same.

This is the order of magnitute that the TCP RFC recommends, but most implementations use a much shorter value. Similarly, the QUIC draft was changed after discussing with the TCP folks to use the same value, but not many implementations are willing to take that performance hit. quic-go retransmits two copies of the Initial packet after 200ms (with exponential backoff after that).

> The real question is whether or not the ICMP unreachable packet this is exposed to the reader. @marten-seemann, any ideas on this?

For UDP, the kernel will only deliver ICMP packets if you're using a connected socket.

You can't really rely on ICMP packets anyway, since they're frequently dropped (or mis-routed). Furthermore, as ICMP packets are not authenticated, it would be inadvisable to take any action, except for the very early stages of the handshake.